### PR TITLE
 Custom Tooltips for Vector Layers

### LINKED ISSUE #27: Add Custom Tooltips for Vector Layers
To mimic: https://w08-mapserver.compass.ie/mapview-examples/examples/customtooltips.html

Example code: https://w08-mapserver.compass.ie/mapview-examples/examples/customtooltips.js

Source code: https://github.com/compassinformatics/mapview/blob/master/src/gis/ToolTips.js

Tooltips configuration is loaded from the layer config file. Original tooltips were created using a custom OL2 class. 

image

In mapview tooltips are configured with the tooltipsconfig object. See https://w08-mapserver.compass.ie/mapview-config-docs/tooltipsConfig.html#tooltipsconfig-json

Properties matched OL vector feature properties, and an optional alias was allowed to have a more "human-friendly" field title.

 ```
           "tooltipsConfig": [
                {
                    "property": "PropertyName1"
                }, 
                {
                    "alias": "Property Name 1", 
                    "property": "PropertyName1"
                }, 
```

### PR DESCRIPTION
**Pull request by chrismayer**
This adds custom tooltips for hovered vector features on the map. The tooltips can be enabled if the corresponding layer has the property `tooltipsConfig` in its JSON-configuration:

grafik

Closes #27.

### CODE REVIEW

File: app/factory/Layer.js [hunk 1]
```diff
@@ -190,12 +190,35 @@ Ext.define('CpsiMapview.factory.Layer', {
...
-            //     https://github.com/meggsimum/cpsi-mapview/issues/27
+            // create a custom toolitp for this layer
+            var toolTip = Ext.create('CpsiMapview.view.layer.ToolTip', {
+                toolTipConfig: layerConf.tooltipsConfig,
+                layer: wfsLayer
+            });
+            wfsLayer.toolTip = toolTip;
+
+            // show / hide on appropriate events
+            mapPanel.on('cmv-map-pointerrest', function(hoveredObjs, evt) {
```
Comment (marcjansen) [inline app/factory/Layer.js]: Do we need to check if the hovered feat is from this layer?

Comment (chrismayer) [reply app/factory/Layer.js]: This is implicitly done in Line 214. The tooltip, which is rendered, is bound to the layer and is only available if the layer is 'hoverable'.

Comment (marcjansen) [reply app/factory/Layer.js]: What I meant is the following: if we duplicate the layer GAS WFS and call it GAS WFS 2, and put in a log in line 2017 which prints out the `hoveredObjs`, you'll see, that we get two logs (expected), but both of them have two features (rather unexpected, right?). A possible (untested) fix could be to check `hoveredObj.layer` against `wfsLayer`. Is this more clear now?

**Comment by marcjansen** [comment]
Two other remarks:

**Popups of two or more layers overlapping**

image

This is what can happen when we hav two layers that are hoverable, not sure if this can be circumvented somehow. It can also of course be, that the smaller tooltipo lies below the bigger one, in which case you do not see it.

**Popups of different layers at the same time**

image

same here, not sure what dto do here and whether this is an actual issue.

Nothing blocking, just sth. we need to think about / document?

File: app/view/layer/ToolTip.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            html = me.layer.formatFunction(feature);
+        } else {
+            html = me.formatFunction(feature);
+        }
+
+        // set HTML content
+        me.setHtml(html);
+
+        // show tooltip near mouse pointer
+        var screenX = evt.originalEvent.x +  me.offsetX;
```
[the remark this answers is not in the document]
Comment (chrismayer) [reply app/view/layer/ToolTip.js]: Done in 51cfef734b52.

File: app/factory/Layer.js (hunk 1 above)
Comment (chrismayer) [reply app/factory/Layer.js]: Done in 51cfef734b52.

**Comment by geographika** [comment]
Looks very nice!

@chrismayer - just a note on the old logic - it was also based on the layer order (or technically z-index), so there would only be one tooltip at a single location. It is simple enough for users to turn off layers if they 
want to focus on a particular vector layer and its tooltips.

**Comment by marcjansen** [comment]
Thanks for the feedback, Seth

**Comment by chrismayer** [comment]
Thanks for your reviews @marcjansen and thanks for the feedback @geographika! 

So this is good to go. I'll merge now.

**Pull request merged**